### PR TITLE
feat: Add Get Token Metrics to GRPC server

### PR DESCRIPTION
**Pull request by siddimore**
**Description**

This PR creates a GRPC method to expose some of the Token Metrics

**TODO**
Add tests

**[Signed commits](../CONTRIBUTING.md#signing-off-on-commits-developer-certificate-of-origin)**
- [x] Yes, I signed my commits.

### CODE REVIEW

File: backend/backend.proto [hunk 1]
```diff
@@ -26,6 +26,17 @@ service Backend {
...
 
   rpc Rerank(RerankRequest) returns (RerankResult) {}
+
+  rpc GetMetrics(MetricsRequest) returns (MetricsResponse);
+}
+
+// Define the empty request
+message MetricsRequest {}
+
+message MetricsResponse {
```
Comment (siddimore) [inline backend/backend.proto:36]: what other metrics are good to be exposed out?

**Comment by mudler** [comment]
mmm I wonder if it would make sense to add a route also via http rest to leverage this, but shouldn't block merging this PR

File: backend/backend.proto (hunk 1 above)
Comment (mudler) [reply backend/backend.proto:36]: Maybe not related to the processing request, but could be useful also to expose infos like how many parallel requests can be served

File: pkg/grpc/client.go [hunk 2]
```diff
@@ -374,3 +374,21 @@ func (c *Client) Rerank(ctx context.Context, in *pb.RerankRequest, opts ...grpc.
 	client := pb.NewBackendClient(conn)
 	return client.Rerank(ctx, in, opts...)
 }
+
+func (c *Client) GetTokenMetrics(ctx context.Context, in *pb.MetricsRequest, opts ...grpc.CallOption) (*pb.MetricsResponse, error) {
+	if !c.parallel {
```
Comment (mudler) [inline pkg/grpc/client.go:379]: Code doesn't look formatted here  - might need a `go fmt`

**Comment by siddimore** [comment]
> mmm I wonder if it would make sense to add a route also via http rest to leverage this, but shouldn't block merging this PR

@mudler yup i already did make those changes just did not add it to this PR because i had to investigate some build issue

**Comment by mudler** [comment]
> > mmm I wonder if it would make sense to add a route also via http rest to leverage this, but shouldn't block merging this PR
> 
> @mudler yup i already did make those changes just did not add it yet because i had to investigate some build issue

gotcha :+1: changes looks good here otherwise!

**Review by mudler — DISMISSED** [review]

**Comment by siddimore** [comment]
> mmm I wonder if it would make sense to add a route also via http rest to leverage this, but shouldn't block merging this PR

@mudler added a route to http endpoints. I will update test and have this PR ready for review this week

File: pkg/grpc/client.go (hunk 2 above)
Comment (sidmore) [reply pkg/grpc/client.go:379]: forgot to fix this

Comment (siddimore) [reply pkg/grpc/client.go:379]: forgot to fix this

**Review by mudler — APPROVED** [review]

File: core/http/endpoints/localai/get_token_metrics.go [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+	"github.com/mudler/LocalAI/pkg/model"
+)
+
+// TokenMetricsEndpoint is an endpoint to get TokensProcessed Per Second for Active SlotID
+//
+//	@Summary	Get TokenMetrics for Active Slot.
+//	@Accept json
+//	@Produce audio/x-wav
+//	@Success	200		{string}	binary				"generated audio/wav file"
```
Comment (mintyleaf) [inline core/http/endpoints/localai/get_token_metrics.go:19]: > binary "generated audio/wav file"?

File: core/http/endpoints/localai/get_token_metrics.go [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+//	@Router		/v1/tokenMetrics [get]
+//	@Router		/tokenMetrics [get]
+func TokenMetricsEndpoint(cl *config.BackendConfigLoader, ml *model.ModelLoader, appConfig *config.ApplicationConfig) func(c *fiber.Ctx) error {
```
Comment (mintyleaf) [inline core/http/endpoints/localai/get_token_metrics.go:22]: That endpoint didn't used anywhere at all

If it gets included - it needs to be called in parallel with running slot to get some output by the logic from modified grpc-server.cpp, but i guess loadModel is just waiting until the slot released producing empty output every time

**Comment by mintyleaf** [comment]
That thing is very unfinished and exposed swagger api documentation makes it even more confusing 

To get it to work every request is need to be stored somewhere with that statistics (we have an id field in openai response)

Or the less intrusive but tacky and unreliable way is to fire that route right after firing inference, where the metrics will wait until the finish and return that additional data

File: backend/cpp/llama/grpc-server.cpp [hunk 5]
```diff
@@ -495,6 +495,16 @@ struct llama_server_context
         }
     }
 
+    llama_client_slot* get_active_slot() {
+        for (llama_client_slot& slot : slots) {
+            // Check if the slot is currently processing
+            if (slot.is_processing()) {
```
Comment (mintyleaf) [inline backend/cpp/llama/grpc-server.cpp:501]: if we replace that with slot.available() - if there was at least one inference we get statistics for the last call

if not - we get a 'Server error error="json: unsupported value: NaN"'

**Comment by mintyleaf** [comment]
@mudler 
we already have all the needed data in backend ready to be packed into the message 

maybe we can introduce some request flag that will add some statistics into the response that will be just an addition to the openai response?

for my goals i need to get the token speed and precise inference run time after each request for multiple nodes which running in federated mode

using tokenMetrics endpoint possibly can be routed to the wrong machine after each request, which is not reliable at all

what do you think about that?

**Pull request merged**